massa-protocol-worker-2/src/handlers/block_handler/retrieval.rs

```diff
@@ -0,0 +1,1280 @@
+                                        "protocol.protocol_worker.process_command.wishlist_delta.end",
+                                        {}
+                                    );
+                                }
+                            }
+                        },
+                        Err(err) => {
+                            println!("Error in retrieval block handler channel command: {:?}", err);
+                            return;
+                        }
+                    }
+                },
+                recv(at(self.next_timer_ask_block)) -> _ => {
+                    if let Err(err) = self.update_ask_block() {
+                        warn!("Error in ask_blocks: {:?}", err);
+                    }
+                }
+            }
+        }
+    }
+
+    /// Network ask the local node for blocks
+    ///
+    /// React on another node asking for blocks information. We can forward the operation ids if
+    /// the foreign node asked for `AskForBlocksInfo::Info` or the full operations if he asked for
+    /// the missing operations in his storage with `AskForBlocksInfo::Operations`
+    ///
+    /// Forward the reply to the network.
+    fn on_asked_for_blocks_received(
+        &mut self,
+        from_peer_id: PeerId,
+        list: Vec<(BlockId, AskForBlocksInfo)>,
+    ) -> Result<(), ProtocolError> {
+        let mut all_blocks_info = vec![];
+        for (hash, info_wanted) in &list {
+            let (header, operations_ids) = match self.storage.read_blocks().get(hash) {
+                Some(signed_block) => (
+                    signed_block.content.header.clone(),
+                    signed_block.content.operations.clone(),
+                ),
+                None => {
+                    // let the node know we don't have the block.
+                    all_blocks_info.push((*hash, BlockInfoReply::NotFound));
+                    continue;
+                }
+            };
+            let block_info = match info_wanted {
+                AskForBlocksInfo::Header => BlockInfoReply::Header(header),
+                AskForBlocksInfo::Info => BlockInfoReply::Info(operations_ids),
+                AskForBlocksInfo::Operations(op_ids) => {
+                    // Mark the node as having the block.
+                    {
+                        let mut cache_write = self.cache.write();
+                        cache_write.insert_blocks_known(
+                            &from_peer_id,
+                            &[*hash],
+                            true,
+                            Instant::now(),
+                        );
+                    }
+                    // Send only the missing operations that are in storage.
+                    let needed_ops = {
+                        let operations = self.storage.read_operations();
+                        operations_ids
+                            .into_iter()
+                            .filter(|id| op_ids.contains(id))
```

Comment:
think we can try to remove this filter and do all check in filter_map

Reply:
I don’t see the change

Reply:
https://github.com/massalabs/massa/pull/3840/files#diff-9bb654cd720a34b029e918b26fb2b2e4ed7f43ba55a3c4d8f93511c18d4470c7R265